packages/components/src/presentation-components/tree/controlled/TreeHooks.ts

```diff
@@ -197,33 +182,18 @@ function useModelSourceUpdateOnIModelHierarchyUpdate(params: {
       let subscription: Subscription | undefined;
       const removeListener = Presentation.presentation.onIModelHierarchyChanged.addListener(
         async (args: IModelHierarchyChangeEventArgs) => {
-          if (args.rulesetId !== dataProvider.rulesetId || args.imodelKey !== dataProvider.imodel.key) {
+          if (args.rulesetId !== rulesetId || args.imodelKey !== dataProviderProps.imodel.key) {
             return;
           }
 
-          const newDataProvider = new PresentationTreeDataProvider({ ...dataProviderProps, ruleset: args.rulesetId });
-          if (args.updateInfo === UPDATE_FULL) {
-            subscription = reloadTree(modelSource.getModel(), newDataProvider, pageSize).subscribe({
-              next: (newModelSource) => setTreeNodeLoaderState((prevState) => ({
-                modelSource: newModelSource,
-                rulesetRegistration: prevState.rulesetRegistration,
-                dataProvider: newDataProvider,
-              })),
-            });
-          } else {
-            const newModelSource = await updateModelSourceAfterIModelChange(
-              modelSource,
-              args.updateInfo,
-              newDataProvider,
-              treeNodeItemCreationProps,
-              renderedItems.current,
-            );
-            setTreeNodeLoaderState((prevState) => ({
+          const dataProvider = new PresentationTreeDataProvider({ ...dataProviderProps, ruleset: rulesetId });
+          subscription = reloadTree(modelSource.getModel(), dataProvider, pageSize, renderedItems.current).subscribe({
+            next: (newModelSource) => setTreeNodeLoaderState((prevState) => ({
               modelSource: newModelSource,
               rulesetRegistration: prevState.rulesetRegistration,
-              dataProvider: newDataProvider,
-            }));
-          }
+              dataProvider,
+            })),
+          });
```

Comment:
looks like this part could be reused by all 3 update-related hooks

Reply:
Extracted same code to separate function